source/code/providers/support/mycnf-scanner.cpp

```diff
@@ -0,0 +1,244 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * ---------------------------------- END OF LICENSE -----------------------------
+*/
+/**
+    \file        mycnf-scanner.cpp
+
+    \brief       Scans mysql configuration to extract authentication information.
+
+    \date        2014-11-05 16:22:00
+
+*/
+/*----------------------------------------------------------------------------*/
+
+#include "mycnf-scanner.h"
+
+namespace MySQL
+{
+    static inline bool GetPathFromDirective(const std::wstring& line, SCXCoreLib::SCXFilePath& path)
+    {
+        std::vector<std::wstring> elements;
+        SCXCoreLib::StrTokenize(line, elements, L" ");
+        if ( elements.size() != 2 )
+        {
+            return false;
+        }
+
+        path.Set(elements[1]);
+        return true;
+    }
+
+    bool MySQL_ConfigurationScanner::LoadCnfFile(
+        const SCXCoreLib::SCXFilePath& file, 
+        MySQL_Authentication& auth,
+        std::set< std::wstring >& visited, 
+        std::map<unsigned int, std::wstring>& ports)
+    {
+        SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L" loading: ", file.Get()));
+        std::pair<std::set< std::wstring >::iterator, bool> ins = visited.insert(file.Get());
+        if (!ins.second)
+        {
+            std::wstringstream ss;
+            ss << L"Already visited " << file.Get() << L". skipping";
+            SCX_LOGWARNING(m_log, ss.str());
+            return true;
+        }
+
+        std::vector< std::wstring > lines;
+        try
+        {
+            SCXCoreLib::SCXStream::NLFs nlfs;
+
+            m_extDepends->ReadAllLinesAsUTF8(file, lines, nlfs);
+        }
+        catch (SCXCoreLib::SCXException & e)
+        {
+            std::wstringstream ss;
+            ss << L"Caught " << e.What() << L" when trying to read " << file.Get() ;
+            SCX_LOGERROR(m_log, ss.str());
+            return false;
+        }
+
+        unsigned int portNum = 0;
+        std::wstring bindAddress;
+        for(std::vector< std::wstring >::const_iterator iter = lines.begin(); iter != lines.end(); ++iter)
+        {
+            std::wstring line( SCXCoreLib::StrTrim(*iter) );
+            SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L" Line: ", line));
+            if (SCXCoreLib::StrIsPrefix(line, L"!includedir"))
+            {
+                SCXCoreLib::SCXFilePath directory;
+                if (GetPathFromDirective(line, directory))
+                {
+                    LoadCnfDirectory(directory, auth, visited, ports);
+                }
+                else
+                {
+                    SCX_LOGWARNING(m_log, SCXCoreLib::StrAppend(L" Ignoring Error in !includedir directive.", line));
```

Comment:
There are errors where we'll stop anyway. If an error occurs where we know we'll do the wrong thing (like ignore an includedir or something), then we shouldn't do a "half" approach, reading some of it and not others. Better to just stop, log the issue, and leave the auth file alone. If we can't run properly, then it's better to just stop rather than to essentially "mangle" the auth file by reading some of the config but not others.